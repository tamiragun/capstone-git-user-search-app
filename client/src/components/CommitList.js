import React from "react";

export function CommitList(props) {
  const commits = props.commits;

  const commitsList = commits.map((commit) => {
    return (
      <li key={commit.id} className="commit-bullet">
        {commit.message} <br></br>({commit.committed_date})
      </li>
    );
  });

  return (
    <div>
      <ol className="commits-list">{commitsList}</ol>
    </div>
  );
}
