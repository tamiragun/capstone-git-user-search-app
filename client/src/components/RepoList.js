import React from "react";
import PropTypes from "prop-types";

//This component takes a prop in the form of an array of repo objects,
//and displays an ordered list of them. It also takes a prop with the
//displayRepo function of the parent (SingleUser --> App) for when the
//button is clicked, and a third props which is the user object, which
//is used as an argument in that displayRepo function.
export function RepoList(props) {
  //Extract the list of repos from the given props
  const repos = props.repos;

  //Event handler which is triggered when the button is clicked, and in
  //turn invokes the displayRepo function which was passed down as a prop.
  const handleClick = (e) => {
    //As arguments, pass the source and username from the user prop, and
    //the value of the button that was clicked
    props.displayRepo(props.user.source, props.user.login, e.target.value);
  };

  //For every repo in the array of repos that was passed in as a prop,
  //display the name and assign a unique key to each.
  const reposList = repos.map((repo) => {
    return (
      <li key={repo.id}>
        <div className="repo-bullet">
          <span>{repo.name}</span>
          <button
            //Assign two values to this, as both are needed by the server API
            //This results in a comma separated string which will be handled
            //by the function it is passed to
            value={[repo.name, repo.id]}
            onClick={handleClick}
            className="select-button view-repo-button"
          >
            View this repo
          </button>
        </div>
      </li>
    );
  });

  return <ol className="repo-list">{reposList}</ol>;
}

//The props are required, and should be an array of commit objects, an object,
//and a function.
RepoList.propTypes = {
  repos: PropTypes.array.isRequired,
  user: PropTypes.object.isRequired,
  displayRepo: PropTypes.func.isRequired,
};
