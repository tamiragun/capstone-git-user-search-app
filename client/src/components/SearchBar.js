import React from "react";
import PropTypes from "prop-types";

//This component takes a prop in the form of a search function, and displays
//a form. When input is entered, this component updates its own state and when
//the form is submitted, is passes the info to the search function of the
//parent (App).
export class SearchBar extends React.Component {
  //Set the state to have the search term as an empty string
  constructor(props) {
    super(props);
    this.state = { searchString: "" };

    //Bind the two functions defined below
    this.handleUserInput = this.handleUserInput.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  //Each time something is typed in the input field, update the
  //state to reflect the latest value.
  handleUserInput(e) {
    this.setState({ searchString: e.target.value });
  }

  //Once the form is submitted, execute the search
  handleSubmit(e) {
    //Prevent the browser from refreshing
    e.preventDefault();
    //Call the function passed as a prop by the parent (App)
    //And pass it the latest state of the searchString.
    this.props.handleSearch(this.state.searchString);
    //Clear the input field by setting the state back to an empty string
    this.setState({ searchString: "" });
  }

  render() {
    return (
      <div>
        {/*Upon form submit, call the handler function defined above */}
        <form onSubmit={this.handleSubmit}>
          <input
            type="text"
            //Set the value to whatever the current state is, to control this input field
            value={this.state.searchString}
            //Every time its content changes, call the change handler to update the state
            onChange={this.handleUserInput}
            className="search-input"
            required
          ></input>
          <button type="submit" className="select-button">
            Search
          </button>
        </form>
      </div>
    );
  }
}

//The prop is required, and should be a function.
SearchBar.propTypes = {
  handleSearch: PropTypes.func.isRequired,
};
