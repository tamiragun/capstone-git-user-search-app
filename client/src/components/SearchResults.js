import React from "react";
import PropTypes from "prop-types";

//This component takes a prop in the form of an array of result objects,
//and displays an unordered list of them. It also takes a prop with the
//displayUser function of the parent (App) for when the button is clicked.
export function SearchResults(props) {
  //Extract the list of repos from the given props
  const results = props.results;

  //Event handler which is triggered when the button is clicked, and in
  //turn invokes the displayUser function which was passed down as a prop.
  const handleClick = (e) => {
    //As argument, pass the the value of the button that was clicked
    props.displayUser(e.target.value);
  };

  //For every search result in the array of search results that was passed in
  //as a prop, display the name, source, and assign a unique key to each.
  const resultsList = results.map((result) => {
    return (
      <li key={result.id} className="result-bullet">
        <p>
          <strong>Username: </strong>
          {result.login} <br></br>
          <strong>Source: </strong>
          {result.source} <br></br>
        </p>
        <button
          //Assign two values to this, as both are needed by the server API
          //This results in a comma separated string which will be handled
          //by the function it is passed to
          value={[result.source, result.login]}
          onClick={handleClick}
          className="select-button"
        >
          View this user
        </button>
      </li>
    );
  });

  return <ul className="results-list">{resultsList}</ul>;
}

//The props are required, and should be an array of commit objects, and object,
//and a function.
SearchResults.propTypes = {
  results: PropTypes.array.isRequired,
  displayUser: PropTypes.func.isRequired,
};
