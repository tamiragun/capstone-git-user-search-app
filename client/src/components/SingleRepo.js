import React from "react";
import { CommitList } from "./CommitList";
import PropTypes from "prop-types";
import { getData } from "../utils/getData.js";

//This component takes a prop in the form of a repo object, and displays
//its details. It also takes a prop with the backToUser function of the
//parent (App) for when the button is clicked
export function SingleRepo(props) {
  //commits will determine whether the commits are displayed vs a loading
  //placeholder, and contain the necessary top 5 commits info
  const [commits, setCommits] = React.useState(null);

  //As soon as the component is rendered, this function will run and fetch
  //the list of commits to display.
  React.useEffect(() => {
    //Nest an async function within useEffect, because useEffect is synchronous
    const fetchData = async () => {
      //Construct the server endpoint url using the props
      const url = `/api/commits?source=${props.repo.source}&user=${props.repo.owner}&reponame=${props.repo.name}&repoid=${props.repo.id}`;
      // Fetch the data
      const result = await getData(url);
      //Update the state to contain the array of commits
      setCommits(result);
    };
    //Call the async function to execute the fetch
    fetchData();
    //Run this effect every time the repo info changes
  }, [props.repo]);

  return (
    <div className="single-repo">
      <h2>{props.repo.name}</h2>
      <div className="TBC">
        <p>
          {props.repo.description}
          <br></br>
          <br></br>
          <strong>First created: </strong>
          {props.repo.created_at}
          <br></br>
          <strong>Last updated: </strong>
          {props.repo.updated_at}
          <br></br>
        </p>
        <h3>Latest 5 commits:</h3>
        {/*If the commits fetch is till pending, display a placeholder */}
        {!commits ? (
          <div className="placeholder">Loading...</div>
        ) : /*If the commits fetch yields zero results, display a message */
        commits.length === 0 ? (
          <div className="placeholder">No commits yet</div>
        ) : (
          /*If the commits fetch yields results, display the commits. Pass 
          on the list of commits from the state. */
          <CommitList commits={commits}></CommitList>
        )}
      </div>
      <div className="TBC">
        {/*No need for a separate event handler, this invokes the function 
          that was passed down as a prop */}
        <button onClick={props.backToUser} className="back-button">
          Go back to user profile
        </button>
      </div>
    </div>
  );
}

//The props are required, and should be a repo object and a function.
SingleRepo.propTypes = {
  repo: PropTypes.object.isRequired,
  backToUser: PropTypes.func.isRequired,
};
