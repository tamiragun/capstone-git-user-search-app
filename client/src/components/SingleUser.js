import React from "react";
import { RepoList } from "./RepoList";
const fetch = require("node-fetch");

export function SingleUser(props) {
  const [repos, setRepos] = React.useState(null);

  React.useEffect(async () => {
    const url = `http://localhost:3001/api/repos?source=${props.user.source}&user=${props.user.login}`;
    try {
      const response = await fetch(url);
      //console.log("Response: ", response, " Type: ", typeof response);
      const jsonResponse = await response.json();
      // console.log(
      //   "jsonResponse: ",
      //   jsonResponse,
      //   " Type: ",
      //   typeof jsonResponse
      // );
      setRepos(jsonResponse);
    } catch (err) {
      console.log(err);
    }
  }, [props.user.login]);

  const handleClick = (e) => {
    props.goBack();
  };

  return (
    <div className="single-user">
      <h3>Selected user</h3>
      <div className="TBC">
        <p>
          <strong>Name: </strong>
          {props.user.login}
          <br></br>
          <strong>Url: </strong>
          {props.user.url}
          <br></br>
        </p>
        {!repos ? (
          "Loading"
        ) : (
          <RepoList
            user={props.user.login}
            repos={repos}
            displayRepo={props.displayRepo}
          ></RepoList>
        )}
      </div>
      <div className="TBC">
        <button onClick={props.goBack}>Go back</button>
      </div>
    </div>
  );
}
