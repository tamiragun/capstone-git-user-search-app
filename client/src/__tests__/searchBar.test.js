import React from "react";
import { SearchBar } from "../components/SearchBar";
import renderer from "react-test-renderer";

test("SearchBar is shown", () => {
  const tree = renderer.create(<SearchBar />).toJSON();
  expect(tree).toMatchSnapshot();
});
