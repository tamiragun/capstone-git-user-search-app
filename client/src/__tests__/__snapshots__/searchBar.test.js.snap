// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar is shown 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      className="search-input"
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
    <button
      className="select-button search-button"
      type="submit"
    >
      Search
    </button>
  </form>
</div>
`;
